Speed up counting of bits in allocator

twpayne opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR tidies up counting of bits in `pkg/registry/core/service/allocator`.

Switching from using [`math/big.Int.Bytes`](https://pkg.go.dev/math/big#Int.Bytes) to [`math/big.Int.Bits`](https://pkg.go.dev/math/big#Int.Bits) has two effects:
* It removes an allocation (there is no longer a need for an intermediate `[]byte`).
* It allows the bits to be counted 64 at a time instead of 8 at a time.

#### Which issue(s) this PR fixes:

NONE

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

twpayne commented:
/assign @smarterclayton

twpayne commented:
/unassign @smarterclayton

aojea commented:
we should have some benchmarks demonstrating this

twpayne commented:
> we should have some benchmarks demonstrating this

The code that this replaces performs the following steps:

1. Allocates a `[]byte` for the words in `n` (a `math/big.Int`).
2. Calls [this function](https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/math/big/nat.go;drc=refs%2Ftags%2Fgo1.17.1;l=1158) to copy n's words (effectively a `[]unit64`) into a `[]byte`.
3. Iterates over each byte, counting the set bits 8 at a time.

The replacement code in this PR performs the following steps:

1. Iterates over n's words counting the set bits 64 at a time.

Is it necessary to demonstrate that the replacement code is faster than the replaced code with a benchmark?

aojea commented:
> Is it necessary to demonstrate that the replacement code is faster than the replaced code with a benchmark?

yep

twpayne committed:
Add extra test and benchmark for count bits

twpayne committed:
Speed up counting of bits in allocator

Benchmark:

goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/registry/core/service/allocator
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz

Before:

BenchmarkCountBits-8     9459236               140.4 ns/op

After:

BenchmarkCountBits-8    140667842                9.541 ns/op

twpayne commented:
OK, benchmark:

Before:
```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/registry/core/service/allocator
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkCountBits-8     9459236               140.4 ns/op
```

After:
```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/registry/core/service/allocator
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkCountBits-8    140667842                9.541 ns/op
```

twpayne commented:
/assign @smarterclayton

aojea commented:
/ok-to-test
/lgtm
/assign @thockin 
naive question, can there be some issues in 32 bits architectures?

twpayne commented:
> naive question, can there be some issues in 32 bits architectures?

On 32-bit machines `math/big.Int`s use `[]uint32`s instead of `[]uint64`, however the [`math/bits.OnesCount64`](https://pkg.go.dev/math/bits#OnesCount64) will still correctly count the bits in a `uint32` as the upper bits will be zero when Go converts the `uint32` to a `uint63`.

Note that `math/bits.OnesCount64` is [already used by Kubernetes](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/bitmask/bitmask.go#L166).

thockin commented on pkg/registry/core/service/allocator/utils.go:
```diff
@@ -24,8 +24,8 @@ import (
 	var count int = 0
-	for _, b := range n.Bytes() {
-		count += bits.OnesCount8(uint8(b))
+	for _, w := range n.Bits() {
```
The docs for math/big.Word say "A Word represents a single digit of a multi-precision unsigned integer" which is certainly not explaining very much.

The docs  for Bits() say "Bits is intended to support implementation of missing low-level Int functionality outside this package; it should be avoided otherwise."   I guess this counts as "missing low-level Int functionality".

thockin reviewed (commented):
Thanks!

/lgtm
/approve

aojea commented:
/sig network
as the allocator belong to services???
